c++/src/capnp/compat/websocket-rpc-test.c++

```diff
@@ -0,0 +1,73 @@
+
+#include "websocket-rpc.h"
+#include <kj/test.h>
+
+#include "json.capnp.h" // Arbitrary schema, just so we have something to play with
+
+namespace capnp::_::WebSocketMessageStream {
+  class FailErrorHandler final : public kj::TaskSet::ErrorHandler {
+    public:
+      void taskFailed(kj::Exception&& exception) override {
+        KJ_FAIL_ASSERT(exception);
+      }
+  };
+};
+
+KJ_TEST("WebSocketMessageStream") {
+  auto io = kj::setupAsyncIo();
+
+  auto pipe = kj::newWebSocketPipe();
+
+  auto msgStreamA = capnp::WebSocketMessageStream(*pipe.ends[0]);
+  auto msgStreamB = capnp::WebSocketMessageStream(*pipe.ends[1]);
+
+  // Make a message, fill it with some stuff
+  capnp::MallocMessageBuilder originalMsg;
+  auto object = originalMsg.initRoot<capnp::json::Value>().initObject(10);
+  object[0].setName("Test");
+  object[0].initValue().setString("A string");
+  object[1].setName("Another field");
+  object[1].initValue().setNumber(42);
+
+  auto originalSegments = originalMsg.getSegmentsForOutput();
+
+  capnp::_::WebSocketMessageStream::FailErrorHandler errorHandler;
+  kj::TaskSet tasks(errorHandler);
+
+  // Send the message across the websocket, make sure it comes out unharmed.
+  tasks.add(msgStreamA.writeMessage(nullptr, originalSegments));
+  tasks.add(msgStreamB.tryReadMessage(nullptr)
+    .then([&](auto maybeResult) -> kj::Promise<void> {
```

Comment:
This is cosmetic, but usually I'd write this like:

```suggestion
  kj::Promise<void> writePromise = msgStreamA.writeMessage(nullptr, originalSegments);
  auto maybeResult = msgStreamB.readMessage(nullptr).wait(waitScope);
  writePromise.wait(waitScope);
```

Then proceed with the checking of the result.

I think it comes out a little easier to read, and you can avoid the need to define `FailErrorHandler` (which, incidentally, as written, might not be correct -- I don't think you can just rethrow the exception in an error handler, since the purpose of the error handler was to handle an uncaught exception).

Reply:
This does appear to work as you suggest, but it wasn't obvious to me from the docs that you can rely on the write end resolving if nothing is waiting on the read end -- the comment for newWebSocketPipe() says:

> No buffering occurs -- a message send does not complete until a corresponding receive accepts the message.

...which I interpreted as suggesting that the version you describe could deadlock. Is this relying on undocumented implementation details?

Really what I want here is to be able to join two promises and reject if either of them rejects, and let the event loop handle the scheduling -- the TaskSet with the error handler that just rethrows was ultimately a verbose way of getting that, since I couldn't find something relevant elsewhere in the library.

Reply:
> it wasn't obvious to me from the docs that you can rely on the write end resolving if nothing is waiting on the read end

The write end won't resolve until a read is performed on the other end. That's why my code creates the write promise but doesn't wait on it, then performs the read, and only after that verifies that the write completed.

> Really what I want here is to be able to join two promises and reject if either of them rejects

Admittedly if the write rejected immediately (without actually writing) then this code would end up deadlocking waiting for the read end, and you'd have to poke the code a bit to figure out what happened. I've been meaning to add a way to join promises with nice syntax and fail-fast behavior... but haven't gotten around to it yet.

If you were really worried about it, you could check `readPromise.poll()` to verify the read actually completes before you `wait()` it -- and if it doesn't complete, then go back and check on `writePromise`. But I don't think it's worth it.